plutus-tx/test/laws/Main.hs

```diff
@@ -0,0 +1,533 @@
+  (x Plutus.+ y) Plutus.* z === (x Plutus.* z) Plutus.+ (y Plutus.* z)
+
+propScaleOne :: Property
+propScaleOne = property $ do
+  x <- forAllWith ppShow genRational
+  Plutus.scale Plutus.one x === x
+
+propScaleDistPlus :: Property
+propScaleDistPlus = property $ do
+  x <- forAllWith ppShow genInteger
+  y <- forAllWith ppShow genRational
+  z <- forAllWith ppShow genRational
+  Plutus.scale x (y Plutus.+ z) === Plutus.scale x y Plutus.+ Plutus.scale x z
+
+propScaleTimes :: Property
+propScaleTimes = property $ do
+  x <- forAllWith ppShow genInteger
+  y <- forAllWith ppShow genInteger
+  r <- forAllWith ppShow genRational
+  Plutus.scale x (Plutus.scale y r) === Plutus.scale (x Plutus.* y) r
+
+propUnsafeIsData :: Property
+propUnsafeIsData = property $ do
+  x <- forAllWith ppShow genRational
+  (unsafeFromBuiltinData . toBuiltinData $ x) === x
+
+propZeroDenom :: Property
+propZeroDenom = property $ do
+  x <- forAllWith ppShow genInteger
+  Ratio.ratio x Plutus.zero === Nothing
+
+propOneDenom :: Property
+propOneDenom = property $ do
+  x <- forAllWith ppShow genInteger
+  Ratio.ratio x Plutus.one === (Just . Ratio.fromInteger $ x)
+
+propConstructionAgreement :: Property
+propConstructionAgreement = property $ do
+  n <- forAllWith ppShow genInteger
+  d <- forAllWith ppShow . Gen.filter (/= 0) $ genInteger
+  Ratio.ratio n d === (Just . Ratio.unsafeRatio n $ d)
+
+propFromIntegerNum :: Property
+propFromIntegerNum = property $ do
+  x <- forAllWith ppShow genInteger
+  let r = Ratio.fromInteger x
+  Ratio.numerator r === x
+
+propFromIntegerDen :: Property
+propFromIntegerDen = property $ do
+  x <- forAllWith ppShow genInteger
+  let r = Ratio.fromInteger x
+  Ratio.denominator r === Plutus.one
+
+propPosDen :: Property
+propPosDen = property $ do
+  r <- forAllWith ppShow genRational
+  assert (Ratio.denominator r Plutus.>= Plutus.one)
+
+propDoubleNeg :: Property
+propDoubleNeg = property $ do
+  r <- forAllWith ppShow genRational
+  (Ratio.negate . Ratio.negate $ r) === r
+
+propAbs :: Property
+propAbs = property $ do
+  r <- forAllWith ppShow genRational
+  assert (Ratio.abs r >= Plutus.zero)
+
+propIsDataRound :: Property
+propIsDataRound = property $ do
+  r <- forAllWith ppShow genRational
+  tripping r toBuiltinData fromBuiltinData
+
+propIsJSONRound :: Property
+propIsJSONRound = property $ do
+  r <- forAllWith ppShow genRational
+  tripping r encode decode
+
+propRatioScale :: Property
+propRatioScale = property $ do
+  x <- forAllWith ppShow genInteger
+  y <- forAllWith ppShow genInteger
+  z <- forAllWith ppShow . Gen.filter (/= Plutus.zero) $ genInteger
+  Ratio.ratio x y === Ratio.ratio (x Plutus.* z) (y Plutus.* z)
+
+propRatioSelf :: Property
+propRatioSelf = property $ do
+  x <- forAllWith ppShow . Gen.filter (/= Plutus.zero) $ genInteger
+  Ratio.ratio x x === Just Plutus.one
+
+propRatioSign :: Property
+propRatioSign = property $ do
+  (n, d) <- forAllWith ppShow go
+  cover 30 "zero numerator" $ n == 0
```

Comment:
> FWIW, I've found that current settings give such failures rarely, but I'm happy to adjust to make it even less likely.

The problem is that any errors in the tests will cause CI to fail, so we might get random failures in completely unrelated PRs because of this, and when that happens someone has to restart the CI job for the PR manually.  If it happened one time in a thousand that _might_ be bearable, but still not ideal.  I just re-ran the tests another 100 times and this time got seven coverage failures, which is definitely too much.

Reply:
I'll do similar experiments locally with more tests and lower coverage boundaries, using 1-in-a-thousand as the desired target. It'll take a while, but I can definitely bring it within that bound.

Reply:
We're definitely not asking you to fix Hedgehog :)